osu.Game.Rulesets.Catch/UI/Catcher.cs

```diff
@@ -464,127 +469,62 @@ private void updateState(CatcherAnimationState state)
             updateCatcher();
         }
 
-        private void placeCaughtObject(DrawablePalpableCatchHitObject drawableObject, Vector2 position)
-        {
-            var caughtObject = getCaughtObject(drawableObject.HitObject);
-
-            if (caughtObject == null) return;
-
-            caughtObject.CopyStateFrom(drawableObject);
-            caughtObject.Anchor = Anchor.TopCentre;
-            caughtObject.Position = position;
-            caughtObject.Scale /= 2;
-
-            caughtObjectContainer.Add(caughtObject);
-
-            if (!caughtObject.StaysOnPlate)
-                removeFromPlate(caughtObject, DroppedObjectAnimation.Explode);
-        }
-
-        private Vector2 computePositionInStack(Vector2 position, float displayRadius)
-        {
-            const float radius_div_2 = CatchHitObject.OBJECT_RADIUS / 2;
-            const float allowance = 10;
-
-            while (caughtObjectContainer.Any(f => Vector2Extensions.Distance(f.Position, position) < (displayRadius + radius_div_2) / (allowance / 2)))
-            {
-                float diff = (displayRadius + radius_div_2) / allowance;
-
-                position.X += (RNG.NextSingle() - 0.5f) * diff * 2;
-                position.Y -= RNG.NextSingle() * diff;
-            }
-
-            position.X = Math.Clamp(position.X, -CatcherArea.CATCHER_SIZE / 2, CatcherArea.CATCHER_SIZE / 2);
-
-            return position;
-        }
-
         private void addLighting(CatchHitObject hitObject, float x, Color4 colour)
         {
             HitExplosion hitExplosion = hitExplosionPool.Get();
-            hitExplosion.HitObject = hitObject;
+            hitExplosion.LifetimeStart = hitObject.StartTime;
             hitExplosion.X = x;
             hitExplosion.Scale = new Vector2(hitObject.Scale);
             hitExplosion.ObjectColour = colour;
             hitExplosionContainer.Add(hitExplosion);
         }
 
-        private CaughtObject getCaughtObject(PalpableCatchHitObject source)
+        private void addCaughtObject(DrawablePalpableCatchHitObject hitObject, double time, Vector2 positionInStack)
         {
-            switch (source)
-            {
-                case Fruit _:
-                    return caughtFruitPool.Get();
+            CaughtObjectEntry entry;
 
-                case Banana _:
-                    return caughtBananaPool.Get();
-
-                case Droplet _:
-                    return caughtDropletPool.Get();
-
-                default:
-                    return null;
+            if (hitObject.HitObject is Droplet)
+            {
+                // droplet explodes immediately
+                entry = new CaughtObjectEntry(CaughtObjectState.Dropped, positionInStack, hitObject)
+                {
+                    ApplyTransforms = applyExplodeTransforms,
+                    DropPosition = CaughtObjectContainer.GetCurrentDropPosition(positionInStack),
```

Comment:
Seems weird for the catcher to have to call the container's own method like this. Should probably be `CaughtObjectContainer`'s responsibility to calculate this (which should hopefully allow `GetCurrentDropPosition` to become private).